File: go/2458. Height of Binary Tree After Subtree Removal Queries.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func treeQueries(root *TreeNode, queries []int) []int {
    nodeLevel := make(map[int]int)     // node  -> level
    levelHeight := make(map[int][]int) // level -> heights 2 nodes
    nodeHeight := make(map[int]int)    // node  -> height
    totalHeight := dfs(root, nodeLevel, nodeHeight, levelHeight, 0)
    // fmt.Println(nodeLevel)
    // fmt.Println(levelHeight)
    // fmt.Println(nodeHeight)
    result := make([]int, len(queries))
    for i, q := range queries {
        level := nodeLevel[q]
        heights := levelHeight[level]
        if heights[1] == 0 {
            result[i] = totalHeight - nodeHeight[heights[0]] - 1
        } else if q != heights[0] && q != heights[1] {
            result[i] = totalHeight
        } else if q == heights[1] {
            result[i] = totalHeight
        } else {
            result[i] = totalHeight - nodeHeight[heights[0]] + nodeHeight[heights[1]]
        }
    }
    return result
}

func dfs(root *TreeNode, nodeLevel, nodeHeight map[int]int, levelHeight map[int][]int, level int) int {
    if root == nil { return 0 }
    if root.Left == nil && root.Right == nil {
        nodeLevel[root.Val] = level
        nodeHeight[root.Val] = 0
        if _, ok := levelHeight[level]; !ok {
            levelHeight[level] = make([]int, 2)
        }
        heights := levelHeight[level]
        if nodeHeight[heights[0]] <= 0 {
            heights[1], heights[0] = heights[0], root.Val
        } else if nodeHeight[heights[1]] <= 0 {
            heights[1] = root.Val
        }
        levelHeight[level] = heights
        return 0
    }

    leftHeight := dfs(root.Left, nodeLevel, nodeHeight, levelHeight, level + 1)
    rightHeight := dfs(root.Right, nodeLevel, nodeHeight, levelHeight, level + 1)
    height := max(leftHeight, rightHeight)+1

    nodeLevel[root.Val] = level
    if _, ok := levelHeight[level]; !ok {
        levelHeight[level] = make([]int, 2)
    }
    heights := levelHeight[level]
    if nodeHeight[heights[0]] < height {
        heights[1], heights[0] = heights[0], root.Val
    } else if nodeHeight[heights[1]] < height {
        heights[1] = root.Val
    }
    levelHeight[level] = heights
    nodeHeight[root.Val] = height
    return height
}

File: java/2458. Height of Binary Tree After Subtree Removal Queries.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int[] treeQueries(TreeNode root, int[] queries) {
        Map<Integer, Integer> subtreeHeight = new HashMap<>();
        Map<Integer, Integer> nodeLevel = new HashMap<>();
        Map<Integer, int[]> levelNodes = new HashMap<>();
        populate(root, subtreeHeight);
        populateBFS(root, subtreeHeight, nodeLevel, levelNodes);
        int totalHeight = subtreeHeight.get(root.val);
        int[] answer = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int node = queries[i];
            int level = nodeLevel.get(node);
            int[] nodes = levelNodes.get(level);
            if (nodes[0] == node) {
                answer[i] = totalHeight - subtreeHeight.get(node)
                    + (nodes[1] != -1 ? subtreeHeight.get(nodes[1]) : -1);
            } else {
                answer[i] = totalHeight;
            }
        }
        return answer;
    }
    private int populate(TreeNode current, Map<Integer, Integer> subtreeHeight) {
        if (current == null) return -1;
        int left = populate(current.left, subtreeHeight);
        int right = populate(current.right, subtreeHeight);
        int height = Math.max(left, right) + 1;
        subtreeHeight.put(current.val, height);
        return height;
    }
    private void populateBFS(
        TreeNode root,
        Map<Integer, Integer> subtreeHeight,
        Map<Integer, Integer> nodeLevel,
        Map<Integer, int[]> levelNodes
    ) {
        Deque<TreeNode> q = new LinkedList<>();
        q.addLast(root);
        int currentLevel = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            int[] nodes = {-1, -1};
            while (size-- > 0) {
                TreeNode current = q.pollFirst();
                nodeLevel.put(current.val, currentLevel);
                int height = subtreeHeight.get(current.val);
                if (nodes[0] == -1 || subtreeHeight.get(nodes[0]) < height) {
                    nodes[1] = nodes[0];
                    nodes[0] = current.val;
                } else if (nodes[1] == -1 || subtreeHeight.get(nodes[1]) < height) {
                    nodes[1] = current.val;
                }
                if (current.left != null) q.addLast(current.left);
                if (current.right != null) q.addLast(current.right);
            }
            levelNodes.put(currentLevel, nodes);
            currentLevel++;
        }
    }
}